secret-memory/src/allocator.rs

```diff
@@ -0,0 +1,239 @@
+    fn allocate(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
+        match &self.inner {
+            AllocInner::SecretAlloc(alloc) => alloc.do_allocate(layout),
+            AllocInner::SodiumAlloc(alloc) => alloc.do_allocate(layout),
+        }
+    }
+
+    unsafe fn deallocate(&self, ptr: NonNull<u8>, layout: Layout) {
+        match &self.inner {
+            AllocInner::SecretAlloc(alloc) => alloc.do_deallocate(ptr, layout),
+            AllocInner::SodiumAlloc(alloc) => alloc.do_deallocate(ptr, layout),
+        }
+    }
+}
+
+impl Default for Alloc {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl fmt::Debug for Alloc {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match &self.inner {
+            AllocInner::SecretAlloc(_) => fmt.write_str("<memfd_secret based Rust allocator>"),
+            AllocInner::SodiumAlloc(_) => fmt.write_str("<memsec based Rust allocator>")
+        }
+    }
+}
+
+#[cfg(not(unix))]
+fn get_support_memfd_secret() -> bool {
+    false
+}
+
+#[cfg(unix)]
+fn get_support_memfd_secret() -> bool {
+    static SUPPORT_MEMFD_SECRET: std::sync::OnceLock<bool> = std::sync::OnceLock::new();
+    *SUPPORT_MEMFD_SECRET.get_or_init(|| {
+        let fd  = unsafe {
+            libc::syscall(libc::SYS_memfd_secret, 0) as i32
+        };
+        if fd != -1 || if let Some(os_err) = Error::last_os_error().raw_os_error() { os_err != 38 } else { true } {
+            // with memfd_secret support
+            return true;
+        }
+        if fd != -1 {
+            unsafe {
+                libc::close(fd);
+            }
+        }
+        false
+    })
+}
+
+#[derive(Clone, Default)]
+struct SodiumAlloc;
+
+impl SodiumAlloc {
+    fn do_allocate(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
+        // Call sodium allocator
+        let ptr = unsafe { memsec::malloc_sized(layout.size()) };
+
+        let ptr = if let Some(p) = ptr {
+            p
+        } else {
+            log::error!(
+                "Allocation {layout:?} was requested but memsec returned a null pointer"
+            );
+            return Err(AllocError);
+        };
+
+        // Ensure the right allocation is used
+        let off = (ptr.as_ptr() as *const libc::c_void).align_offset(layout.align());
+        if off != 0 {
+            log::error!("Allocation {layout:?} was requested but memsec returned allocation \
+                with offset {off} from the requested alignment. Memsec always allocates values \
+                at the end of a memory page for security reasons, custom alignments are not supported. \
+                You could try allocating an oversized value.");
+            return Err(AllocError);
+        }
+
+        Ok(ptr)
+    }
+
+    fn do_deallocate(&self, ptr: NonNull<u8>, _layout: Layout) {
+        unsafe {
+            memsec::free(ptr)
+        }
+    }
+}
+
+
+#[derive(Clone)]
+struct SecretAlloc {
+    fd : i32,
```

Comment:
This will always allocate *the same secret* so this is not a functioning allocator.